chain/src/chain.rs

```diff
@@ -111,116 +116,94 @@ impl<CI: ChainIndex + 'static> ChainService<CI> {
+    fn insert_block(&self, block: &Block) -> Result<BlockInsertionResult, SharedError> {
+        let mut new_best_block = false;
+        let mut total_difficulty = U256::zero();
 
-        for tx in b.commit_transactions() {
-            let ins = if tx.is_cellbase() {
-                Vec::new()
-            } else {
-                tx.input_pts()
-            };
-            let outs = tx.output_pts();
+        let mut old_blocks = Vec::new();
+        let mut new_blocks = Vec::new();
 
-            cells.push((ins, outs));
-        }
+        let mut txo_set_diff = TxoSetDiff::default();
 
-        let root = self
+        let mut chain_state = self.shared.chain_state().write();
+        let tip_number = chain_state.tip_number();
+        let parent_ext = self
             .shared
-            .output_root(&b.header().parent_hash())
-            .ok_or(SharedError::InvalidOutput)?;
-
-        self.shared
             .store()
-            .update_transaction_meta(batch, root, cells)
-            .ok_or(SharedError::InvalidOutput)
-    }
-    #[allow(clippy::op_ref)]
-    fn insert_block(&self, block: &Block) -> Result<BlockInsertionResult, SharedError> {
-        let mut new_best_block = false;
-        let mut output_root = H256::zero();
-        let mut total_difficulty = U256::zero();
+            .get_block_ext(&block.header().parent_hash())
+            .expect("parent already store");
 
-        let mut old_cumulative_blks = Vec::new();
-        let mut new_cumulative_blks = Vec::new();
+        let cannon_total_difficulty = parent_ext.total_difficulty + block.header().difficulty();
+        let current_total_difficulty = chain_state.total_difficulty();
 
-        let mut tip_header = self.shared.tip_header().write();
-        let tip_number = tip_header.number();
-        self.shared.store().save_with_batch(|batch| {
-            let root = self.check_transactions(batch, block)?;
-            let parent_ext = self
-                .shared
-                .store()
-                .get_block_ext(&block.header().parent_hash())
-                .expect("parent already store");
-            let cannon_total_difficulty = parent_ext.total_difficulty + block.header().difficulty();
-
-            let ext = BlockExt {
-                received_at: unix_time_as_millis(),
-                total_difficulty: cannon_total_difficulty.clone(),
-                total_uncles_count: parent_ext.total_uncles_count + block.uncles().len() as u64,
-            };
+        debug!(
+            target: "chain",
+            "difficulty current = {}, cannon = {}",
+            current_total_difficulty,
+            cannon_total_difficulty,
+        );
 
+        let ext = BlockExt {
+            received_at: unix_time_as_millis(),
+            total_difficulty: cannon_total_difficulty.clone(),
+            total_uncles_count: parent_ext.total_uncles_count + block.uncles().len() as u64,
+            valid: if parent_ext.valid == Some(false) {
```

Comment:
Why continue processing the new block here if its parent block validation result is false?

Reply:
we should save it

Reply:
We can just save this ext and quit.